web-ts/src/routes/CityContentSwitcher.tsx

```diff
@@ -0,0 +1,118 @@
+import React, { ReactElement, useCallback } from 'react'
+import { Route, RouteComponentProps, Switch, generatePath } from 'react-router-dom'
+import OffersPage from './offers/OffersPage'
+import EventsPage from './events/EventsPage'
+import CategoriesPage from './categories/CategoriesPage'
+import PoisPage from './pois/PoisPage'
+import NewsPage from './news/NewsPage'
+import SearchPage from './search/SearchPage'
+import DisclaimerPage from './disclaimer/DisclaimerPage'
+import {
+  EVENTS_ROUTE,
+  OFFERS_ROUTE,
+  POIS_ROUTE,
+  LOCAL_NEWS_TYPE,
+  TU_NEWS_TYPE,
+  DISCLAIMER_ROUTE,
+  SEARCH_ROUTE,
+  CityModel,
+  useLoadFromEndpoint,
+  createLanguagesEndpoint,
+  LanguageModel,
+  CATEGORIES_ROUTE,
+  LANDING_ROUTE
+} from 'api-client'
+import { cmsApiBaseUrl } from '../constants/urls'
+import Layout from '../components/Layout'
+import FailureSwitcher from '../components/FailureSwitcher'
+import { LanguageFailure } from '../components/LanguageFailure'
+import GeneralHeader from '../components/GeneralHeader'
+import GeneralFooter from '../components/GeneralFooter'
+import LocationLayout from '../components/LocationLayout'
+import LoadingSpinner from '../components/LoadingSpinner'
+import { cityContentPattern, RoutePatterns } from './RootSwitcher'
+
+type PropsType = {
+  cities: CityModel[]
+} & RouteComponentProps<{ cityCode: string; languageCode: string }>
+
+// TODO pass right props instead of constants
+const CityContentSwitcher = ({ cities, match, location }: PropsType): ReactElement => {
+  const { cityCode, languageCode } = match.params
+  const cityModel = cities.find(it => it.code === cityCode)
+
+  const requestLanguages = useCallback(async () => {
+    return createLanguagesEndpoint(cmsApiBaseUrl).request({ city: cityCode })
+  }, [cityCode])
+  const { data: languages, loading, error: languagesError } = useLoadFromEndpoint<LanguageModel[]>(requestLanguages)
+  const languageModel = languages?.find(it => it.code === languageCode)
```

Comment:
Loading and finding of cities and languages are entirely handled here (and in `RootSwitcher`), so no more nasty checks in all routes.

Reply:
I think we should not handle this in a centralized way. I think this code is likely to change with new requirements.

Right now we do things this way:

```
<Router>
<SomeLayout>
<RouteXY>
</RouteXY>
</SomeLayout>
</Router>
```

I think this is the wrong way around. It should be:

```
<Router>
<RouteXY>
<SomeLayout>
</SomeLayout>
</RouteXY>
</Router>
```

This is because "The route determines the layout" and not the other way around: The layout determines the route.

The nasty checks you mentioned come from the fact that we used the state and called the functions which fill the state from a centralized code in the app. If you do this, then you have to check in every components whether the state is "ready".

So I propose to create a handful of reusable hooks like: `useCity`, `useCategories`, `useEvents`. They are called in each route. Until they are not finished display a loading spinner. As soon as they are finished display the content or an error. The type of layout rendering is also happen in each route. This will lead to some duplicate, but very simple code. Also then you see the dependencies of each route very clearly. Like CategoriesRoute requires cities, page and events and therefore sends 3 fetch requests.

If we add the new MapsView for the maps feature which requires a special Layout then there is no way of implementing it without touching this file. That should not be neccassary.

Reply:
So I think you are actually mentioning two issues here @maxammann, which are somehow related though.
1) Where the loading happens, i.e. whether we load cities and languages in the switcher and pass those to the routes or load it separately in each route:
I actually don't really see a problem with my approach here, dependencies are clear as those are passed as props and there is just no need to reload this whenever a user navigates. The fetching of the specific content like categories, events and so on should definitely happen in the routes though and not in that central place.

2) Whether the layout is wrapped around the route or the route around the layout:
The problem with rendering the app (as @Taggotty mentioned) was the reason why I chose to implement it this way, otherwise the layout is umounted and remounted whenever one navigates, which will give us the feeling of a simple old website imo.
New specific layouts for new routes aren't a real problem either as we can just not wrap it in the LocationLayout and render a layout in the route. Not very nice to handle layouts in different places but I think not to rerender the layout all the time is more important. Or do you see another way to prevent this?

Reply:
Other opinions?

Reply:
So actually rerendering is not a problem, react detects that it is the same component automatically and keeps it rendered. Therefore done in #310.